### PR TITLE
Make Gloo an optional c10d dependency

### LINKED ISSUE #18851: Allow torch.distributed to be built/used without Gloo
See `torch/csrc/distributed/c10d/init.cpp` -- it unconditionally includes `ProcessGroupGloo.hpp`.

Together with fixing this, also update the `is_gloo_available` function (see #18595).

### PR DESCRIPTION
**Pull request by pietern**
The CMake modifications include removal of some unnecessary paths
(e.g. find_package(CUDA) and friends) that are no longer used since
c10d is always part of the larger torch build. The macro
`C10D_USE_...` was ambiguous and is now removed in favor of only
having top level `USE_...`. The c10d test suite is changed to include
skip annotations for the tests that depend on Gloo as well.

Now, if you compile with `USE_DISTRIBUTED=1` and `USE_GLOO=0` you get
a functioning build for which the tests actually pass.

Closes #18851.

### CODE REVIEW

File: torch/lib/c10d/CMakeLists.txt [hunk 1]
```diff
@@ -7,32 +7,29 @@ list(APPEND CMAKE_MODULE_PATH
...
+if(USE_GLOO)
+  option(USE_C10D_GLOO "USE C10D GLOO" ON)
+endif()
+
+if(USE_NCCL)
+  option(USE_C10D_NCCL "USE C10D NCCL" ON)
 endif()
 
 find_package(MPI)
 if(MPI_FOUND)
```
Comment (mrshenli) [inline torch/lib/c10d/CMakeLists.txt]: Curious, why there is no `USE_MPI` option? Does it mean MPI is enabled as long as the lib is found?

File: torch/lib/c10d/CMakeLists.txt [hunk 2]
```diff
@@ -7,32 +7,29 @@ list(APPEND CMAKE_MODULE_PATH
...
-  set(C10D_USE_CUDA false)
-  message(STATUS "Building C10D without CUDA support")
+  message(STATUS "Building c10d without CUDA/ROCm support")
+endif()
+
+if(USE_GLOO)
+  option(USE_C10D_GLOO "USE C10D GLOO" ON)
+endif()
+
+if(USE_NCCL)
```
Comment (mrshenli) [inline torch/lib/c10d/CMakeLists.txt]: what if `USE_NCCL` is ON but `USE_CUDA` is off?

**Review by mrshenli — APPROVED** [review]

File: torch/lib/c10d/CMakeLists.txt (hunk 1 above)
Comment (pietern) [reply torch/lib/c10d/CMakeLists.txt]: Good point. Wrapped this block and made it conditional on USE_MPI.

File: torch/lib/c10d/CMakeLists.txt (hunk 2 above)
Comment (pietern) [reply torch/lib/c10d/CMakeLists.txt]: This case is handled earlier by `cmake/Dependencies.cmake` and `USE_NCCL` is set to `OFF`.

**Pull request closed without merging**